### PR TITLE
Support low latency wg

### PR DESCRIPTION
**Pull request by laotan332**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

### CODE REVIEW

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 1]
```diff
@@ -8,11 +8,16 @@
...
-GlobalDriverDispatcher::GlobalDriverDispatcher(std::unique_ptr<ThreadPool> thread_pool)
-        : _driver_queue(std::make_unique<DriverQueueWithWorkGroup>()),
-          _thread_pool(std::move(thread_pool)),
+GlobalDriverDispatcher::GlobalDriverDispatcher(std::unique_ptr<ThreadPool> thread_pool, bool is_low_latency)
+        : _thread_pool(std::move(thread_pool)),
           _blocked_driver_poller(new PipelineDriverPoller(_driver_queue.get())),
-          _exec_state_reporter(new ExecStateReporter()) {}
+          _exec_state_reporter(new ExecStateReporter()) {
+    if (is_low_latency) {
+        _driver_queue = std::make_unique<QuerySharedDriverQueue>();
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:16]: `QuerySharedDriverQueue` in the workgroup_main branch is different from that in the main branch.
Maybe we should rename `QuerySharedDriverQueue` in the workgroup_main branch to something like `DriverQueueWithoutLock` and copy ``QuerySharedDriverQueue` in the main branch to workgroup_main.

File: be/src/exec/workgroup/work_group.cpp [hunk 2]
```diff
@@ -265,6 +265,34 @@ StatusOr<PriorityThreadPool::Task> IoWorkGroupQueue::pick_next_task() {
...
+    }
+
+    return _real_time_wg->pick_io_task();
+}
+
+StatusOr<PriorityThreadPool::Task> WorkGroupManager::pick_next_task_for_io(bool is_real_time_type) {
+    if (is_real_time_type) {
+        return _wg_io_queue.pick_next_real_time_task();
+    }
+    return _wg_io_queue.pick_next_task();
```
Comment (ZiheLiu) [inline be/src/exec/workgroup/work_group.cpp:289]: If we implement two different io `WorkGroupQueue` for real-time and non-real-time, and make `io_dispatcher` own `WorkGroupQueue`, then maybe we needn't add method `pick_next_real_time_task` and `pick_next_ask`, and needn't judge whether is real-time here.

File: be/src/exec/pipeline/scan_operator.cpp [hunk 3]
```diff
@@ -168,8 +169,16 @@ Status ScanOperator::_trigger_next_scan(RuntimeState* state, int chunk_source_in
     // TODO(by satanson): set a proper priority
     task.priority = 20;
 
-    if (WorkGroupManager::instance()->try_offer_io_task(_workgroup, task)) {
-        _io_task_retry_cnt = 0;
+    if (_workgroup != nullptr) {
+        if (_workgroup->type() == WorkGroupType::WG_REALTIME) {
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/scan_operator.cpp:173]: Maybe we could add a method like `WorkGroup::is_realtime`.

File: be/src/exec/pipeline/pipeline_driver_queue.h [hunk 4]
```diff
@@ -30,37 +46,66 @@ class DriverQueue {
     virtual void put_back_from_dispatcher(const std::vector<DriverRawPtr>& drivers) = 0;
 
     virtual StatusOr<DriverRawPtr> take() = 0;
+    virtual StatusOr<DriverRawPtr> take(size_t* queue_index) = 0;
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver_queue.h:49]: Maybe we needn't this interface, and make `QuerySharedDriverQueue::put_back()` record `queue_index` in `PipelineDriver::_dispatch_queue_index` like `DriverQueueWithouLock::put_back()`.

File: be/src/exec/pipeline/pipeline_driver_queue.h [hunk 5]
```diff
@@ -30,37 +46,66 @@ class DriverQueue {
...
 
     // Update statistics of the driver's workgroup,
     // when yielding the driver in the dispatcher thread.
     virtual void update_statistics(const DriverRawPtr driver) = 0;
 
+    virtual SubQuerySharedDriverQueue* get_sub_queue(size_t) = 0;
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver_queue.h:55]: Maybe we also needn't this interface.

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 6]
```diff
@@ -87,8 +116,11 @@ void GlobalDriverDispatcher::run() {
             // query context has ready drivers to run, so extend its lifetime.
             query_ctx->extend_lifetime();
             auto status = driver->process(runtime_state);
-            this->_driver_queue->update_statistics(driver);
-
+            if constexpr (is_use_workgroup) {
+                this->_driver_queue->update_statistics(driver);
+            } else {
+                this->_driver_queue->get_sub_queue(queue_index)->update_accu_time(driver);
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:122]: We could update statistics in `QuerySharedDriverQueue::update_statistics`, and then we needn't judge whether `is_use_workgroup` here..

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 7]
```diff
@@ -107,7 +139,11 @@ void GlobalDriverDispatcher::run() {
             switch (driver_state) {
             case READY:
             case RUNNING: {
-                this->_driver_queue->put_back_from_dispatcher(driver);
+                if constexpr (is_use_workgroup) {
+                    this->_driver_queue->put_back_from_dispatcher(driver);
+                } else {
+                    this->_driver_queue->put_back(driver);
+                }
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:146]: we could make `QuerySharedDriverQueue::put_back_from_dispatcher()` call `QuerySharedDriverQueue::put_back()`, and then we needn't judge whether `is_use_workgroup` here.

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 8]
```diff
@@ -8,11 +8,17 @@
...
+        : _thread_pool(std::move(thread_pool)),
           _blocked_driver_poller(new PipelineDriverPoller(_driver_queue.get())),
-          _exec_state_reporter(new ExecStateReporter()) {}
+          _exec_state_reporter(new ExecStateReporter()),
+          _is_use_workgroup(is_use_workgroup) {
+    if (_is_use_workgroup) {
+        _driver_queue = std::make_unique<QuerySharedDriverQueue>();
+    } else {
+        _driver_queue = std::make_unique<DriverQueueWithWorkGroup>();
+    }
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:20]: Maybe `_is_use_workgroup==true` should use `DriverQueueWithWorkGroup`, and `_is_use_workgroup==false` should use `QuerySharedDriverQueue`. Or maybe `_is_use_workgroup` should be renamed to `_is_realtime`.

File: be/src/exec/workgroup/work_group.cpp [hunk 9]
```diff
@@ -278,15 +306,28 @@ bool IoWorkGroupQueue::try_offer_io_task(WorkGroupPtr wg, const PriorityThreadPo
...
-    if (_is_closed) {
-        return;
+        _is_closed = true;
+        _cv.notify_all();
     }
 
-    _is_closed = true;
-    _cv.notify_all();
+    {
+        std::unique_lock<std::mutex> lock(_realtime_io_mutex);
```
Comment (ZiheLiu) [inline be/src/exec/workgroup/work_group.cpp:328]: Check `_is_closed` and set `_is_closed = true` after `lock`.

File: be/src/exec/pipeline/pipeline_driver_queue.h [hunk 10]
```diff
@@ -18,6 +18,22 @@ namespace pipeline {
 class DriverQueue;
 using DriverQueuePtr = std::unique_ptr<DriverQueue>;
 
+class SubQuerySharedDriverQueue {
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver_queue.h:21]: What's the meaning of `SubQuery` ?

File: be/src/exec/workgroup/work_group.cpp (hunk 2 above)
Comment (kangkaisen) [reply be/src/exec/workgroup/work_group.cpp:432]: +1

**Comment by ZiheLiu** [comment]
run starrocks_fe_unittest

File: be/src/runtime/exec_env.cpp [hunk 11]
```diff
@@ -167,6 +169,31 @@ Status ExecEnv::_init(const std::vector<StorePath>& store_paths) {
     _io_dispatcher = new workgroup::IoDispatcher(std::move(io_dispatcher_thread_pool));
     _io_dispatcher->initialize(num_io_threads);
 
+    // for real time
+    std::unique_ptr<ThreadPool> real_time_driver_dispatcher_thread_pool;
```
Comment (kangkaisen) [inline be/src/runtime/exec_env.cpp:173]: do we need create real_time_driver_dispatcher_thread_pool and real_time_io_dispatcher_thread_pool by default？

File: be/src/runtime/exec_env.h [hunk 12]
```diff
@@ -130,6 +130,7 @@ class ExecEnv {
     PriorityThreadPool* etl_thread_pool() { return _etl_thread_pool; }
     FragmentMgr* fragment_mgr() { return _fragment_mgr; }
     starrocks::pipeline::DriverDispatcher* driver_dispatcher() { return _driver_dispatcher; }
+    starrocks::pipeline::DriverDispatcher* real_time_driver_dispatcher() { return _real_time_driver_dispatcher; }
```
Comment (kangkaisen) [inline be/src/runtime/exec_env.h:133]: comment why we need a real_time_driver_dispatcher?

File: be/src/exec/pipeline/pipeline_driver_queue.h (hunk 10 above)
Comment (laotan332) [reply be/src/exec/pipeline/pipeline_driver_queue.h:21]: it be used for Multi-level Feedback Queue

File: be/src/runtime/exec_env.cpp (hunk 11 above)
Comment (laotan332) [reply be/src/runtime/exec_env.cpp:173]: Yes, we need to, and a system will only allow the creation of one

**Pull request closed without merging**